Allow set pred with ConstantVariable

guangy10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98909
* __->__ #98900

It's part of the effort to improve PT2 Export UX. This PR is to improve the usability of `torch.cond()` by allowing user to set `pred` as `ConstantVariable` as it's not often to see control flow on rank or a tensor or dim size which is traced as `ConstantVariable`.


cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

guangy10 committed:
Allow set pred with ConstantVariable

[ghstack-poisoned]

jansel commented on test/functorch/test_control_flow.py:
```diff
@@ -378,6 +378,35 @@ def forward(self, y_1):
+            return cond(pred, true_fn, false_fn, [x])
+
+        def f_pred_symnode_var(x):
+            from torch._dynamo.variables.tensor import SymNodeVariable
+            pred: SymNodeVariable = x.shape[0] > 10
```
```suggestion
            pred = x.shape[0] > 10
```

jansel commented on test/functorch/test_control_flow.py:
```diff
@@ -378,6 +378,35 @@ def forward(self, y_1):
+            return cond(pred, true_fn, false_fn, [x])
+
+        def f_pred_tensor_var(x):
+            from torch._dynamo.variables.tensor import TensorVariable
+            pred: TensorVariable = x.all()
```
```suggestion
            pred = x.all()
```

jansel reviewed (changes requested):
Type annotations should refer to python types not torchdynamo types.

guangy10 committed:
Update on "Allow set pred with ConstantVariable"


It's part of the effort to improve PT2 Export UX. This PR is to improve the usability of `torch.cond()` by allowing user to set `pred` as `ConstantVariable` as it's not often to see control flow on rank or a tensor or dim size which is traced as `ConstantVariable`.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

guangy10 commented:
@pytorchmergebot merge
